fraud/service.go

```diff
@@ -0,0 +1,147 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+const (
+	fetchHeaderTimeout = time.Minute * 2
+)
+
+// service is responsible for validating and propagating Fraud Proofs.
+// It implements the Service interface.
+type service struct {
+	topicsLk sync.RWMutex
+	topics   map[ProofType]*topic
+
+	pubsub *pubsub.PubSub
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		pubsub: p,
+		getter: getter,
+		topics: make(map[ProofType]*topic),
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.topicsLk.Lock()
+	t, ok := f.topics[proofType]
+	f.topicsLk.Unlock()
+	if !ok {
+		return nil, errors.New("fraud: unmarshaler for the given proof type is not registered")
+	}
+	return newSubscription(t)
+}
+
+func (f *service) RegisterUnmarshaler(proofType ProofType, u ProofUnmarshaler) error {
+	f.topicsLk.Lock()
+	defer f.topicsLk.Unlock()
+
+	t, err := f.pubsub.Join(getSubTopic(proofType))
+	if err != nil {
+		return err
+	}
+	err = f.pubsub.RegisterTopicValidator(
+		getSubTopic(proofType),
+		func(ctx context.Context, _ peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+			return f.processIncoming(ctx, proofType, msg)
+		},
+		// make validation synchronous.
+		pubsub.WithValidatorInline(true),
+	)
+	if err != nil {
+		return err
+	}
+
+	f.topics[proofType] = &topic{topic: t, unmarshal: u}
```

Comment:
Should we check if the unmarshaller is already registered?

Reply:
Or allow overwriting

Reply:
We don't call it on init and its runtime or build time for the Node. We can also log this out. Fine with this, as long as it gives something to debug if someone additionally register it twice